libyul/optimiser/StackToMemoryMover.h

```diff
@@ -0,0 +1,100 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Optimisation stage that moves Yul variables from stack to memory.
+ */
+
+#pragma once
+
+#include <libyul/optimiser/ASTWalker.h>
+#include <libyul/optimiser/OptimiserStep.h>
+#include <libsolutil/Common.h>
+
+namespace solidity::yul
+{
+
+/**
+ * Optimisation stage that moves Yul variables from stack to memory.
+ * It takes a map from functions names and variable names to memory offsets.
+ * It then transforms the AST as follows:
+ *
+ * Single variable declarations are replaced by mstore's as follows:
+ *   If a is in the map, replace
+ *     let a
+ *   by
+ *     mstore(<memory offset for a>, 0)
+ *   respectively, replace
+ *     let a := expr
+ *   by
+ *     mstore(<memory offset for a>, expr)
+ *
+ * In a multi-variable declaration, variables to be moved are replaced by fresh variables and then moved to memory:
```

Comment:
Do we really need the distinction here? A single run of the ExressionJoiner should do the trick, shouldn't it?

Reply:
I think you originally requested that I distinguish them ;-). I'm not sure running the expression joiner (or *any* other optimization step) after the memory escalator is a good idea - but even without doing that, the multi-variable case would work just as well for the single-variable case due to the stack optimization in codegen (and hopefully with peephole optimization actually the exact same code will be generated)...

Reply:
Actually removing the special handling hardly saves any code or complexity... I'll push it as a separate commit, but I actually think it would be nicer to keep the distinction, given that it hardly simplifies the code not to do it.